File: 프로그래머스/2/17684. ［3차］ 압축/［3차］ 압축.java
import java.util.*;

class Solution {
    
    Map<String, Integer> lzwTable;
    
    public int[] solution(String msg) {
        List<Integer> answer = new ArrayList<>();
        lzwTable = initTable();
        
        int lastValue = 26;
        
        for (int i = 0; i < msg.length(); i++) {
            int index = i;
            String check = msg.substring(index, index + 1);
            while(lzwTable.containsKey(check) && index < msg.length() - 1) {
                index++;
                i++;
                check += msg.substring(index, index + 1);
            }
            if (check.length() > 1) {
                if (index == msg.length() - 1 && lzwTable.containsKey(check)) {
                    answer.add(lzwTable.get(check));
                    continue;
                }
                lzwTable.put(check, ++lastValue);
                answer.add(lzwTable.get(check.substring(0, check.length() - 1)));
                i--;
            } else {
                answer.add(lzwTable.get(check));
            }
        }
        return listToArray(answer);
    }
    
    private Map<String, Integer> initTable() {
        Map<String, Integer> lzwTable = new HashMap<>();
        int A = 65;
        for (int i = 0; i < 26; i++) {
            lzwTable.put(String.valueOf((char) (65 + i)), i + 1);
        }
        
        return lzwTable;
    }
    
    private int[] listToArray(List<Integer> input) {
        int[] answer = new int[input.size()];
        
        for (int i = 0; i < input.size(); i++) {
            answer[i] = input.get(i);
        }
        
        return answer;
    }
}

File: 프로그래머스/2/17684. ［3차］ 압축/［3차］ 압축.py
# LZW 압축 구현하기
# 현재 위치에 사전에 포함된 가장 긴 길의 문자열을 구하고 추가한다
# 사전에 없는 문자열의 경우 
def solution(msg):
    answer = []
    lzw = {"A":1, "B":2, "C":3, "D":4, "E":5, "F":6, "G":7, "H":8, "I":9, "J":10, "K":11, "L":12, "M":13, "N":14, "O":15, "P":16, "Q":17, "R":18, "S":19, "T":20, "U":21, "V":22, "W":23, "X":24, "Y":25, "Z":26}
    
    startIndex = 0
    lastNumber = 27
    while startIndex < len(msg):
        string = msg[startIndex]
        
        notFind = False
        # 사전이 있는 가장 긴 문자열 찾기
        for i in range(startIndex + 1, len(msg)):
            string += msg[i]
            if string not in lzw:
                startIndex = i
                notFind = True
                break
            
        
        if notFind:
            lzw[string] = lastNumber
            lastNumber += 1
            answer.append(lzw[string[:-1]])
        else:
            answer.append(lzw[string])
            startIndex = len(msg)
            
    return answer